src/Application/Common/Behaviours/DomainEventPublishingBehavior.cs

```diff
@@ -0,0 +1,46 @@
+﻿using CleanArchitecture.Application.Common.Interfaces;
+using CleanArchitecture.Domain.Common;
+using MediatR.Pipeline;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.Extensions.Logging;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CleanArchitecture.Application.Common.Behaviours
+{
+    public class DomainEventPublishingBehaviour<TRequest, TResponse> : IRequestPostProcessor<TRequest, TResponse>
+    {
+        private readonly ILogger _logger;
+        private readonly IApplicationDbContext _dbContext;
+        private readonly IDomainEventService _domainEventService;
+
+        public DomainEventPublishingBehaviour(
+            ILogger<TRequest> logger,
+            IApplicationDbContext dbContext,
+            IDomainEventService domainEventService)
+        {
+            _logger = logger;
+            _dbContext = dbContext;
+            _domainEventService = domainEventService;
+        }
+
+        public async Task Process(TRequest request, TResponse response, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                var domainEventEntity = _dbContext.ChangeTracker.Entries<IHasDomainEvent>()
+                                                     .Select(x => x.Entity.DomainEvents)
+                                                     .SelectMany(x => x)
+                                                     .Where(domainEvent => !domainEvent.IsPublished)
+                                                     .FirstOrDefault();
+                if (domainEventEntity == null) break;
+
+                domainEventEntity.IsPublished = true;
+                await _domainEventService.Publish(domainEventEntity);
+                
+                _logger.LogInformation("Published event: {Name}", domainEventEntity.GetType().Name);
+            }
+        }
```

Comment:
Using an infinity loop looks like not the best solution. Just my preference not to use `while-true loops` if it's possible to avoid them.

For example, this code is entirely like yours but without infinity looks.

```cs
var domainEventEntities = _dbContext.ChangeTracker.Entries<IHasDomainEvent>()
                                  .Select(x => x.Entity.DomainEvents)
                                  .SelectMany(x => x)
                                  .Where(domainEvent => !domainEvent.IsPublished);
                    
if (domainEventEntities is null)
	return;
              
foreach(var @event in domainEventEntities)
{
	domainEventEntity.IsPublished = true;
    await _domainEventService.Publish(domainEventEntity);

	_logger.LogInformation("Published event: {Name}", domainEventEntity.GetType().Name);
}
```